crud/request.go

```diff
@@ -0,0 +1,1129 @@
+package crud
+
+import (
+	"context"
+
+	"github.com/tarantool/go-tarantool"
+)
+
+type baseRequest struct {
+	impl *tarantool.CallRequest
+}
+
+// Code returns IPROTO code for CRUD request.
+func (req *baseRequest) Code() int32 {
+	return req.impl.Code()
+}
+
+// Ctx returns a context of CRUD request.
+func (req *baseRequest) Ctx() context.Context {
+	return req.impl.Ctx()
+}
+
+// Async returns is CRUD request expects a response.
+func (req *baseRequest) Async() bool {
+	return req.impl.Async()
+}
+
+type spaceRequest struct {
+	baseRequest
+	space interface{}
+}
+
+func (req *spaceRequest) setSpace(space interface{}) {
+	req.space = space
+}
+
+// InsertRequest helps you to create request
+// object to call `crud.insert` for execution
+// by a Connection.
+type InsertRequest struct {
+	spaceRequest
+	tuple interface{}
+	opts  interface{}
+}
+
+// NewInsertRequest returns a new empty InsertRequest.
+func NewInsertRequest(space interface{}) *InsertRequest {
+	req := new(InsertRequest)
+	req.impl = tarantool.NewCallRequest("crud.insert")
+	req.setSpace(space)
+	req.tuple = []interface{}{}
+	req.opts = map[string]interface{}{}
+	return req
+}
+
+// Tuple sets the tuple for the InsertRequest request.
+// Note: default value is nil.
+func (req *InsertRequest) Tuple(tuple interface{}) *InsertRequest {
+	req.tuple = tuple
+	return req
+}
+
+// Opts sets the options for the insert request.
+// Note: default value is nil.
+func (req *InsertRequest) Opts(opts interface{}) *InsertRequest {
+	req.opts = opts
+	return req
+}
```

Comment:
To begin with, opts are always optional. In Lua, it's okay to not specify options, to pass `nil` as options, to pass `{}` as option and specify only those options you'll need (and in this case. each `opts` field which is not filled will be equal to `nil`). Default struct in Go behaves different -- if field is a number, it's initialized with `0`, if field is a string, it's initialized with `''`. For crud, option with value `0` mostly not treated as default -- `0` is used as value, same as option with value `''`. So it is the one issue we will need to inspect thoroughly.

Second, it would become a bit more bothersome to support crud: you'll need to add each new option here when it's introduced in `crud` (for now, it is only an issue with new methods). Moreover, there are several different opts structures depending on what method is used. And in case of methods with the same options, there is no way to guarantee that they will not use different structures later.

On the other hand, it's easy to see the appeal of static here.

So, at first glance it seems like a bother.

Reply:
PR #266